### PR TITLE
Section on backporting

### PR DESCRIPTION
**Pull request by JesperIRL**
Guidance on backporting.
Also moved "Backing out a change" to the testing section where I believe it fits better. This section is unchanged.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - Author) ⚠️ Review applies to afbb243214b48fed3f6c27bfac0dcca994d2036c
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - no project role) ⚠️ Review applies to d5f7a7fcd1c3ff922bb46733f48bdf48ea843dc4
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - Author) ⚠️ Review applies to d5f7a7fcd1c3ff922bb46733f48bdf48ea843dc4
 * [Christoph Langer](https://openjdk.java.net/census#clanger) (@RealCLanger - no project role) ⚠️ Review applies to ab420f92afa37f70e36e08d6f23a17023416e1a7
 * [Goetz Lindenmaier](https://openjdk.java.net/census#goetz) (@GoeLin - no project role) ⚠️ Review applies to 9779db8ad5ff35078805b7bd6106b609d3ec3760
 * [Iris Clark](https://openjdk.java.net/census#iris) (@irisclark - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/guide pull/66/head:pull/66` \
`$ git checkout pull/66`

Update a local copy of the PR: \
`$ git checkout pull/66` \
`$ git pull https://git.openjdk.java.net/guide pull/66/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 66`

View PR using the GUI difftool: \
`$ git pr show -t 66`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/guide/pull/66.diff">https://git.openjdk.java.net/guide/pull/66.diff</a>

</details>

### CODE REVIEW

File: src/index.md
Comment (dholmes-ora) [inline src/index.md:1450]: typo: repostiory

**Review by dholmes-ora — APPROVED** [review]
Hi Jesper,

Changes seem reasonable to me.

Thanks,
David

File: src/index.md
Comment (kevinrushforth) [inline src/index.md:1492]: You might want to add a sentence saying that if a backport record is created ahead of time, all work done in JBS is still done in the main issue.

File: src/index.md
Comment (irisclark) [inline src/index.md:1377]: Recommend that both instances of "reviewer*" in the final sentence be of the same case.  I would choose upper.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Thank you for reviewing! Fixed the typo.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Thank you for reviewing! Added the sentence.

File: src/index.md
Comment (JesperIRL) [reply src/index.md:1377]: The first "reviewers" does not refer to two Reviewers. For instance in Hotspot two reviewers are required but only one of them have to be a Reviewer.

Comment (irisclark) [reply src/index.md:1377]: Ah! Thanks for the clarification.  The existing case make sense.

**Review by kevinrushforth — APPROVED** [review]

File: src/index.md
Comment (stefank) [inline src/index.md:1500]: "wait for the maintainers of the release to approve your request" - Should this section describe that they do so by adding the label `jdk<release>u-fix-yes`?

File: src/index.md
Comment (magicus) [inline src/index.md:1492]:
```suggestion
There can be cases where it is motivated to create a backport issue before the fix is done. In these cases set the fix version of the backport to `N-pool`, where `N` is the release train the backport is targeting. E.g. `17-pool`. Please note that even if a backport issue is created ahead of time, all work done in JBS is still done in the main issue.
```

File: src/index.md
Comment (magicus) [inline src/index.md:1504]:
```suggestion
The Skara tooling includes support for backports. [The official Skara documentation](https://wiki.openjdk.java.net/display/SKARA/Backports) describes in detail how to work with the tooling to create backport PRs on GitHub or using the CLI tools. As described in the documentation, the [`/backport`](https://wiki.openjdk.java.net/display/SKARA/Commit+Commands#CommitCommands-/backport) command can be used on a commit (not a PR!) to create the backport PR. If a backport PR is manually created, set the title to `Backport <original commit hash>`.
```

**Review by magicus — APPROVED** [review]
Looks good in general. I fixed a typo for you. :) Also recommended some changes wrt Skara.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Yes. Thank you for reviewing! Added a sentence with the label.

File: src/index.md
Comment (theRealAph) [inline src/index.md:1375]:
```suggestion
If a change causes a regression that can't be fixed within reasonable time, the best way to handle the regression can be to back out the change. Backing out means that the inverse (anti-delta) of the change is pushed to effectively undo the change in the repository. There are two parts to this task, how to do the bookkeeping in JBS, and how to do the actual backout in git or mercurial.
```

File: src/index.md
Comment (magicus) [reply src/index.md]: You did not like these suggestions? It has happened more than once that users try to use `/backport` on a PR, since commit commands (or even comments in general!) is not widely known about. I think at least that should be made more clear.

Comment (JesperIRL) [reply src/index.md]: These changes has been included in the current version. Thank you for your review and these suggestions!

Comment (magicus) [reply src/index.md]: I see that now. Thanks. It looked to me that they had not been included, but I must have just gotten lost in the Github UI.

**Review by RealCLanger — APPROVED** [review]
Looks good.

File: src/index.md
Comment (GoeLin) [inline src/index.md:1501]: Hi,
I would propose to differ the development and release
repositories most update projects use (like jdk11u-dev and jdk11u).

In this case, in the first sentence, it should say 
"... OpenJDK Update **Development** repositories ..."

and at the end you could add
  If an update is in rampdown in the Release repository, 
  urgent fixes need to be approved with
  `jdk<release>u-critical-request` / `jdk<release>u-critical-yes`.

If you want to keep it short, omitting this is ok.  But this is 
also missing on the "JDK Updates: Requesting push approval for fixes"
wiki page referred to from here.

**Review by GoeLin — CHANGES_REQUESTED** [review]
A general remark wrt. backouts:
One thing I have always been wondering about: 
What to do if a backported issue is backed out?
For an example see JDK-8274338. It was backported
to 11.0.14 but then backed out again as JDK-8276943.
Nevertheless in JBS the original issue is marked as backported 
to 11.0.14.
Should we advise to remove this link, and change
the backport issue and the backout issue to 
bugs?
I think this would make obvious that the issue is NOT 
fixed in the older release (here 11.0.14).

File: src/index.md
Comment (jerboaa) [inline src/index.md:1504]: Suggestion: "... set the PR title to `Backport <original commit hash>`. This ensures that the the bots will recognize it as a backport as opposed to a fix specifically targetting an older release only. One can tell whether or not the bots recognized a PR as a backport by the `backport` label being added if it is being recognized."

File: src/index.md
Comment (gnu-andrew) [inline src/index.md:1494]: Using the exact release e.g. `openjdk8u322` is the only option with 8u.

File: src/index.md
Comment (gnu-andrew) [reply src/index.md:1501]: +1

File: src/index.md
Comment (gnu-andrew) [reply src/index.md:1504]: +1

**Review by gnu-andrew — CHANGES_REQUESTED** [review]

**Comment by JesperIRL** [comment]
> A general remark wrt. backouts:
> One thing I have always been wondering about:
> What to do if a backported issue is backed out?
> For an example see JDK-8274338. It was backported
> to 11.0.14 but then backed out again as JDK-8276943.
> Nevertheless in JBS the original issue is marked as backported
> to 11.0.14.
> Should we advise to remove this link, and change
> the backport issue and the backout issue to
> bugs?
> I think this would make obvious that the issue is NOT
> fixed in the older release (here 11.0.14).

This is an excellent question. I'll add some info on that topic to the section covering backouts.

**Comment by JesperIRL** [comment]
> Note there is a distinction, in this context, between "backing out" a change and later reversing it, or changing it. "backouts" are when a change is pushed and quickly demonstrates it has issues and needs to be undone promptly. But sometimes a change is available for a while and we only later discover it is the source of an unexpected consequence (functional or performance) and so it needs to be modified or sometimes completely reversed. Generally speaking the former should never be backported because a backport should only happen after a change has had some bake time.

In the particular case mentioned here, JDK-8274338, the issue itself was not backed out from mainline, but only the 11.0.14 backport was backed out from 11u. This is a rare case (I hope) but clearly motivates some guidance on how to handle it as the standard model for backports is not sufficient here.

> I would think we should follow what is done for the primary issue in such a case - the backport issue should be Resolved as "Fix failed", and link to the backout issue of course.

I have done that for this backport but this is not enough. The main issue still have the backported-by-link to JDK-8276785 (11.0.14) so in the main issue it still looks like this fix was backported to 11.0.14. A minimum action here would be to also remove the backported-by-link and add a relates-to-link instead.

File: src/index.md
Comment (JesperIRL) [reply src/index.md:1503]: Thank you for these suggestions! I have included this.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Nice catch! Fixed.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Thank you for this suggestion. Included.

File: src/index.md
Comment (JesperIRL) [reply src/index.md]: Thank you for the review. I have included the change and CSR example so resolving this comment. Reply about n-pool in the next comment :-)

File: src/index.md
Comment (JesperIRL) [reply src/index.md:1494]: I don't see why there should be a difference between 8u and later releases. The tooling handle these releases exactly the same way as far as I know. Is there any official communication saying that 8u shouldn't use 8-pool?

**Comment by GoeLin** [comment]
Hi,

Yes, what happened is what Jesper describes.  A change that is good in head
and seems to be fine in an older release still can have problems not noticed
during standard testing.  I think so far we saw this twice in 11u.
Besides introducing a bug in the old release, it might also turn out
that the fix changes the behavior in an undesired way, e.g. breaking the standard.

Yes, replacing the backported-by-link by relates-to is what I would propose, too.
But should we change the type of issue from "Backport" to something  else, probably "Bug"?

And yes, it must be put into this documentation in a way that it is clear that this only applies to backouts of backported issues.

Anyways, thanks for picking this up!

File: src/index.md
Comment (GoeLin) [reply src/index.md:1503]: +1 :)

**Comment by kevinrushforth** [comment]
Are you talking about changing the backport relationship of a backport record in JBS for which a fix has already been pushed, but then subsequently backed out? That seems like the wrong thing to do, given that there is in fact a commit in the update release repo with the main bug ID.

I might be missing something, but would it be sufficient to indicate fix-failed in the backport record (for the release where it was later backed out), and then file a REDO bug?

**Comment by JesperIRL** [comment]
If it's desired to redo the backport filing a REDO bug might be sufficient, but if it's found to be too complex to fix whatever issues caused the backout a REDO bug won't be filed, leaving us in a situation where the main issue looks like it has been backported to 11u when in fact the change is not available in 11u.

Even if the REDO bug is filed, keeping the backported-by-link to the backed out backport would cause confusion as the main issue still looks like it's already backported. And if the second backport is pushed to the same update release there will be two backport issues with the same fix version. In this case JBS will only show one of them in the list of backports, and it's unlikely to be the correct one.

**Comment by JesperIRL** [comment]
I don't think we should change the type of the backed out backport. I don't see any benefit in that, just extra work. We aren't trying to hide the fact that the backport was attempted.

**Comment by kevinrushforth** [comment]
> If it's desired to redo the backport filing a REDO bug might be sufficient, but if it's found to be too complex to fix whatever issues caused the backout a REDO bug won't be filed, leaving us in a situation where the main issue looks like it has been backported to 11u when in fact the change is not available in 11u.

That's a good point: Without a redo bug it will be confusing to anyone looking at it. It's also might be a little confusing to remove the record, but perhaps not as confusing as leaving it there.

> Even if the REDO bug is filed, keeping the backported-by-link to the backed out backport would cause confusion as the main issue still looks like it's already backported. And if the second backport is pushed to the same update release there will be two backport issues with the same fix version.

The redo will use a different main bug IDs, so I don't see that this adds confusion. Your earlier point is probably enough of a reason to do what you suggest.

Btw, I wouldn't suggest deleting the backport relationship for a bug that has been backed out in mainline (shortly) after it was backported, and thus backed out in the update release as well. What you propose makes sense only for bugs that remain in mainline but not in one or more of the update releases. This should be very rare for the reasons you mentioned earlier, but it can happen.

**Comment by JesperIRL** [comment]
> The redo will use a different main bug IDs, so I don't see that this adds confusion. Your earlier point is probably enough of a reason to do what you suggest.

I would expect the redo issue in this case to be a new backport issue from the same main issue.

> Btw, I wouldn't suggest deleting the backport relationship for a bug that has been backed out in mainline (shortly) after it was backported, and thus backed out in the update release as well. What you propose makes sense only for bugs that remain in mainline but not in one or more of the update releases. This should be very rare for the reasons you mentioned earlier, but it can happen.

Right, I totally agree.

**Comment by kevinrushforth** [comment]
>> The redo will use a different main bug IDs, so I don't see that this adds confusion. Your earlier point is probably enough of a reason to do what you suggest.

> I would expect the redo issue in this case to be a new backport issue from the same main issue

That violates our best practices of not reusing a bug. This would lead to two commits in the repo with the same bug ID and different content. There's a reason we don't ever do this in the mainline.

**Comment by JesperIRL** [comment]
> That violates our best practices of not reusing a bug. This would lead to two commits in the repo with the same bug ID and different content. There's a reason we don't ever do this in the mainline.

Right.. So what is the best solution here then?

1. Create the redo issue of type Bug and push it as a separate fix to 11u. Link the new issue with the old main issue as related. Pro: New bugid is a Bug. Cons: The old main issue won't show the new fix as a backport and won't have 11u in the list of backports.
2. Create the redo issue of type Bug and push it as a separate fix to 11u. Link the new issue as a backport of the old main issue. Pro: New bugid is a Bug, old main issue will show the new bug as a backport. Cons: Backport-link to an issue of type Bug.
3. Create the redo issue of type Backport and push it with the redo bugid. Pro: Everything will look good in JBS. Cons: Bugid in change history will refer to a backport issue.
4. Create a new mainline Bug issue and close it as duplicate of the old main issue. Create the redo issue as a backport of the duplicate and push with the duplicate bugid. Pro: Most things will probably look ok, won't break any rules I think. Cons: Seems like  a lot of process for the sake of process, not a very pragmatic solution.

I'd vote for #3 or some better solution not listed here.

**Comment by GoeLin** [comment]
> I don't think we should change the type of the backed out backport issue. I don't see any benefit in that, just extra work. We aren't trying to hide the fact that the backport was attempted.

ok.

**Comment by GoeLin** [comment]
> Btw, I wouldn't suggest deleting the backport relationship for a bug that has been backed out in mainline (shortly) after it was backported, and thus backed out in the update release as well. What you propose makes sense only for bugs that remain in mainline but not in one or more of the update releases. This should be very rare for the reasons you mentioned earlier, but it can happen.
Yes. This issue is not a problem, as the backout would be backported, too, and then again an eventual redo.

**Comment by GoeLin** [comment]
> Right.. So what is the best solution here then?

Is having twice the same bugId such a big problem?
The bugid anyways does not refer to the JBS issue connected to the push.
And these backouts are very rare.

I would propose to remove the backport link and then just do a new backport.  This is the most simple approach for the developers working on the backports. Would this work with skara? Would the repeated bugId break any tooling?

**Comment by kevinrushforth** [comment]
I prefer option 1, but Jesper is right about the con of doing that.

Option 2 seems messy, although it might work.

Option 3 won't work since Skara forbids using a backport ID in a pull request (and I would not be in favor of changing that).

Option 4 seems too heavy-weight and error prone.

> I would propose to remove the backport link and then just do a new backport. This is the most simple approach for the developers working on the backports. Would this work with skara? Would the repeated bugId break any tooling?

Let's call this option 5 then. This won't break any tooling, and from a JBS point of view, won't be confusing. While I still don't much like the idea of having two commits in a repo with the same commit message but with different contents, this seems the most expedient approach. And as noted this should be very rare.

**Comment by GoeLin** [comment]
I think JBS should make transparent in which older release a bug was actually fixed by having a proper backported relation.

> However, those cases should be "the same" fix
We could redo the broken fix as-is with the same bug-id and have the backported relation. Then we could push the adaption to the target release as a follow-up bug.

**Comment by GoeLin** [comment]
Hi, 
sorry, this moved out of my mind.
This seems to be a complicated issue, I don't want to block this PR with it.
Maybe it should be a follow up.
As all the other content is very useful, so we should get it commited.

**Review by GoeLin — APPROVED** [review]
LGTM

**Review by irisclark — APPROVED** [review]

**Pull request closed without merging**
